source/extensions/filters/http/common/utility.h

```diff
@@ -0,0 +1,98 @@
+#pragma once
+
+#include <unordered_map>
+
+#include "envoy/runtime/runtime.h"
+
+#include "common/common/logger.h"
+#include "common/common/macros.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Common {
+
+/**
+ * FilterNameUtil provides utilities for handling HTTP filter
+ * extension names (e.g., "envoy.filters.http.buffer").
+ */
+class FilterNameUtil {
+public:
+  /**
+   * Given a deprecated HTTP filter extension name, return the
+   * canonical name. Any name not defined in the deprecated map is
+   * returned without modification.
+   */
+  static const std::string&
+  canonicalFilterName(const std::string& name,
+                      Runtime::Loader* runtime = Runtime::LoaderSingleton::getExisting()) {
+    const auto& it = deprecatedNameMap().find(name);
+    if (it == deprecatedNameMap().end()) {
+      return name;
+    }
+
+    std::string err = fmt::format(
+        "Using deprecated http filter name '{}' for '{}'. This name will be removed from Envoy "
+        "soon. Please see "
+        "https://www.envoyproxy.io/docs/envoy/latest/intro/deprecated for details.",
+        name, it->second);
+
+#ifdef ENVOY_DISABLE_DEPRECATED_FEATURES
+    bool warn_only = false;
+#else
+    bool warn_only = true;
+#endif
+
+    if (runtime && !runtime->snapshot().deprecatedFeatureEnabled(
+                       "envoy.deprecated_features.allow_historic_http_filter_names", true)) {
+      warn_only = false;
+    }
+
+    if (warn_only) {
+      ENVOY_LOG_MISC(warn, "{}", err);
```

Comment:
Does this also increment the deprecated feature stat? Also, do we need to do a similar thing for other deprecations? Also, is it in any way possible to share the deprecated warn/stat/etc. logic in a helper function that maybe takes an error message, runtime key, etc. since I think we do this pattern in multiple places and it would be useful to consolidate?

Reply:
It does increment the stat (though this is only used for per-filter-configs). I think the registry fallback to the deprecated names only logs. I can take a look at that. 

The other place that invokes deprecatedFeatureEnabled is in protobuf/utility.cc and it has a more complicated flow related to protobuf field deprecations.

The only rough spot that I know of is w.r.t to dynamic metadata. A couple filters whose name changed (namely mongo_proxy and zookeeper) emit metadata using the filter name as the metadata namespace. The metadata seems to be meant for use with rbac/ext_authz/loggers/tracing. I added more information in the deprecated doc, but I couldn't come up with a way to handle it transparently. Or rather, I could probably use an inverse of the deprecated name map to dual-write to both namespaces but logging/stats on use are a mess since there are so many ways it gets accessed.

Reply:
OK whatever you think is best here. If there is no further action let me know and I can take another pass.

Reply:
I think it's ready to go now -- the latest version checks the runtime flag and emits stats from the FactoryRegistry when a deprecated name is used. The check code is shared with the per-filter-config path. For the dynamic metadata I think I'll send something to envoy-users@ when this PR lands.